### PR TITLE
[Merged by Bors] - feat(Algebra/Exact): iff lemmas

### PR DESCRIPTION
**Pull request by joelriou**
When we have a commutative diagram from a sequence of two maps to another, such that the left vertical map is surjective, the middle vertical map is bijective and the right vertical map is injective, then the upper row is exact iff the lower row is. This generalizes the case when all the three vertical maps are bijections. This PR introduces two new lemmas `exact_iff_of_surjective_of_bijective_of_injective` in the `AddMonoidHom` and `LinearMap` namespaces, and refactors the proofs of the all-bijections versions.

This parallels the categorical lemma https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Homology/ShortComplex/Exact.html#CategoryTheory.ShortComplex.exact_iff_of_epi_of_isIso_of_mono

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Exact.lean [hunk 1]
```diff
@@ -90,6 +90,42 @@ lemma exact_of_comp_of_mem_range
     (h1 : g.comp f = 0) (h2 : ∀ x, g x = 0 → x ∈ range f) : Exact f g :=
   exact_of_comp_eq_zero_of_ker_le_range h1 h2
 
+/-- When we have a commutative diagram from a sequence of two maps to another,
+such that the left map is surjective, the middle map is bijective and the right
+map is injective, then the upper row is exact iff the lower row is. -/
```
Comment (chrisflav) [inline Mathlib/Algebra/Exact.lean:95]:
```suggestion
such that the left vertical map is surjective, the middle vertical map is bijective and the right vertical
map is injective, then the upper row is exact iff the lower row is. -/
```

File: Mathlib/Algebra/Exact.lean [hunk 2]
```diff
@@ -396,6 +412,24 @@ end Function
 
 namespace LinearMap
 
+/-- When we have a commutative diagram from a sequence of two linear maps to another,
+such that the left map is surjective, the middle map is bijective and the right
+map is injective, then the upper row is exact iff the lower row is. -/
```
Comment (chrisflav) [inline Mathlib/Algebra/Exact.lean:417]:
```suggestion
/-- When we have a commutative diagram from a sequence of two linear maps to another,
such that the left vertical map is surjective, the middle vertical map is bijective and the right
vertical map is injective, then the upper row is exact iff the lower row is. -/
```

File: Mathlib/Algebra/Exact.lean (hunk 1 above)
Comment (chrisflav) [reply Mathlib/Algebra/Exact.lean:95]: Maybe also link to `CategoryTheory.ShortComplex.exact_iff_of_epi_of_isIso_of_mono` here? (and below)

**Review by chrisflav — APPROVED** [review]

File: Mathlib/Algebra/Exact.lean [hunk 3]
```diff
@@ -90,6 +90,42 @@ lemma exact_of_comp_of_mem_range
     (h1 : g.comp f = 0) (h2 : ∀ x, g x = 0 → x ∈ range f) : Exact f g :=
   exact_of_comp_eq_zero_of_ker_le_range h1 h2
 
+/-- When we have a commutative diagram from a sequence of two maps to another,
+such that the left vertical map is surjective, the middle vertical map is bijective and the right vertical
+map is injective, then the upper row is exact iff the lower row is. -/
```
Comment (joelriou) [inline Mathlib/Algebra/Exact.lean:95]:
```suggestion
map is injective, then the upper row is exact iff the lower row is.
See `ShortComplex.exact_iff_of_epi_of_isIso_of_mono` in the file
`Algebra.Homology.ShortComplex.Exact` for the categorical version of this result. -/
```

**Comment by joelriou** [comment]
Thanks @chrisflav for the reviews!

File: Mathlib/Algebra/Exact.lean [hunk 4]
```diff
@@ -90,6 +90,44 @@ lemma exact_of_comp_of_mem_range
     (h1 : g.comp f = 0) (h2 : ∀ x, g x = 0 → x ∈ range f) : Exact f g :=
   exact_of_comp_eq_zero_of_ker_le_range h1 h2
 
+/-- When we have a commutative diagram from a sequence of two maps to another,
+such that the left vertical map is surjective, the middle vertical map is bijective and the right vertical
+map is injective, then the upper row is exact iff the lower row is.
```
Comment (chrisflav) [inline Mathlib/Algebra/Exact.lean:95]:
```suggestion
such that the left vertical map is surjective, the middle vertical map is bijective and the right
vertical map is injective, then the upper row is exact iff the lower row is.
```
sorry, made the line too long

**Comment by chrisflav** [comment]
Thanks!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**